### ISSUE TITLE
Build TensorFlow Lite for Android and reduce binary size to use with C API 

### ISSUE THREAD

**Issue by cristianl1251** [tensorflow/tensorflow#50946]
Hello, 

I would like to build TensorFlow Lite for Android reducing the binary size. I want to optimize its size based on the operations used by my model. Then, I want to use the generated `.so` for `arm64-v8a` and `armeabi-v7a` in JNI with C API.

I am able to build TensorFlow Lite for Android as described in [Build Android](https://www.tensorflow.org/lite/guide/build_android#build_and_install) and use it in my Android Studio project. However, the binary files are too big for my application. In order to reduce the binary size, I followed the instructions described in [Reduce Binary Size](https://www.tensorflow.org/lite/guide/reduce_binary_size), but as stated in [Known Limitations](https://www.tensorflow.org/lite/guide/reduce_binary_size#known_issueslimitations) it does not support C API. The generated `.aar` does not contain the C headers and when I try to use the `.so` files in my Android Studio project, none of the C functions are available.

Is there any other way to reduce TensorFlow Lite binary size to be used in JNI with C API?

Thanks.

**Comment by abattery** [comment]
Bazel can do the selective build mechanism for C API.

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/c/BUILD#L115

**Comment by abattery** [comment]
The tflite_custom_c_library build rule can generate the selectively built TensorFlow Lite C library.

**Comment by cristianl1251** [comment]
Cool, I replaced `tflite_custom_c_library` build rule with

```
tflite_custom_c_library(
    name = "selectively_built_c_api_test_lib",
    testonly = 1,
    models = [
        "//tensorflow/lite:testdata/mymodel1.bin",
        "//tensorflow/lite:testdata/mymodel2.bin",
    ],
    visibility = ["//visibility:private"],
)
```

Then, built `armeabi-v7a` with

```
bazel build -c opt --config=android_arm //tensorflow/lite/c:selectively_built_c_api_test_lib
```

and `arm64-v8a` with

```
bazel build -c opt --config=android_arm64 //tensorflow/lite/c:selectively_built_c_api_test_lib
```

Tested the generated `.so` files with my app in Android Studio and it built successfully. 

Binaries reduction:

- `arm64-v8a`: 2.5 MB to 1.5 MB
- `armeabi-v7a`: 1.9 MB to 1.7 MB

Is there any reason why binary for `armeabi-v7a` reduces less than `arm64-v8a`?

Thanks for pointing me to the right direction @abattery.

**Comment by abattery** [comment]
I haven't looked at the details. Looks like the pruned op kernels consume more spaces at `arm64-8va`.

**Comment by cristianl1251** [comment]
Hey, I still couldn't make it work. 

When I build TensorFlow Lite using selective build to generate the library and use the library with JNI, my app in Android Studio builds. However, when I flash it to the mobile it cannot load the library. For selective build I use:

```
bazel build -c opt --config=android_arm64 //tensorflow/lite/c:selectively_built_c_api_test_lib
```

In order to verify whether the issue was related to the selective build or not, I also tried building without selective build. Again, I could build my app in Android Studio using the generated library, however it did not work when I flashed it to the mobile. For default build, I use:

```
bazel build -c opt --config=android_arm64 //tensorflow/lite/c:tensorflowlite_c
```

Before, I used to build the library using `tensorflow/lite/tools/build_aar.sh` and I could use the generated library in my app successfully. Successful build:

```
./tensorflow/lite/tools/build_aar.sh --target_archs=arm64-v8a
```

The issue is that, as stated in [Reduce Binary Size](https://www.tensorflow.org/lite/guide/reduce_binary_size#known_issueslimitations), selective build is not supported for C API and I need to reduce my binary size.

I have the following questions: what is the difference between using `bazel build -c opt --config=android_arm64 //tensorflow/lite/c:tensorflowlite_c` and `./tensorflow/lite/tools/build_aar.sh --target_archs=arm64-v8a`? Am I missing anything in `tensorflowlite_c` build rule in order to make it behave like `./tensorflow/lite/tools/build_aar.sh`?

Thanks!
